File: pubsub/src/main/java/glitch/pubsub/object/adapters/VideoPlaybackTypeAdapter.java
package glitch.pubsub.object.adapters;

import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;
import glitch.pubsub.events.json.VideoPlayback;
import java.io.IOException;

public class VideoPlaybackTypeAdapter extends TypeAdapter<VideoPlayback.Type> {

    /**
     * {@inheritDoc}
     */
    @Override
    public void write(JsonWriter out, VideoPlayback.Type value) throws IOException {
        out.value(value.name().toLowerCase());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public VideoPlayback.Type read(JsonReader in) throws IOException {
        String type = in.nextString().replace("-", "_");
        if (type.equalsIgnoreCase("viewcount")) {
            return VideoPlayback.Type.VIEW_COUNT;
        } else {
            return VideoPlayback.Type.valueOf(type.toUpperCase());
        }
    }
}

File: pubsub/src/main/kotlin/io/glitchlib/pubsub/model/adapters/VideoPlaybackTypeAdapter.kt
package io.glitchlib.pubsub.model.adapters

import com.google.gson.TypeAdapter
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonWriter
import io.glitchlib.pubsub.model.json.VideoPlayback
import java.io.IOException

class VideoPlaybackTypeAdapter : TypeAdapter<VideoPlayback.Type>() {

    /**
     * {@inheritDoc}
     */
    @Throws(IOException::class)
    override fun write(out: JsonWriter, value: VideoPlayback.Type) {
        out.value(value.name.toLowerCase())
    }

    /**
     * {@inheritDoc}
     */
    @Throws(IOException::class)
    override fun read(`in`: JsonReader): VideoPlayback.Type {
        val type = `in`.nextString().replace("-", "_")
        return if (type.equals("viewcount", ignoreCase = true)) {
            VideoPlayback.Type.VIEW_COUNT
        } else {
            VideoPlayback.Type.valueOf(type.toUpperCase())
        }
    }
}
